Firmware/MotorControl/pwm_input.cpp

```diff
@@ -0,0 +1,423 @@
+static uint32_t all_counts[PWM_IN_COUNT];
+
+/* Function prototypes -------------------------------------------------------*/
+
+void pwm_no_pulse_check_cb(void *argument);
+
+/* Function implementations --------------------------------------------------*/
+
+// @brief Returns the ODrive GPIO number for a given
+// TIM2 or TIM5 input capture channel number.
+int tim_2_5_channel_num_to_gpio_num(int channel) {
+    #if HW_VERSION_MAJOR == 3 && HW_VERSION_MINOR >= 3
+        if (channel >= 1 && channel <= 4) {
+            // the channel numbers just happen to coincide with
+            // the GPIO numbers
+            return channel;
+        } else {
+            return -1;
+        }
+    #else
+        // Only ch4 is available on v3.2
+        if (channel == 4) {
+            return 4;
+        } else {
+            return -1;
+        }
+    #endif
+}
+
+// @brief Returns the TIM2 or TIM5 channel number
+// for a given GPIO number.
+uint32_t gpio_num_to_tim_2_5_channel(int gpio_num) {
+    #if HW_VERSION_MAJOR == 3 && HW_VERSION_MINOR >= 3
+        switch (gpio_num) {
+            case 1: return TIM_CHANNEL_1;
+            case 2: return TIM_CHANNEL_2;
+            case 3: return TIM_CHANNEL_3;
+            case 4: return TIM_CHANNEL_4;
+            default: return 0;
+        }
+    #else
+        // Only ch4 is available on v3.2
+        if (gpio_num == 4) {
+            return TIM_CHANNEL_4;
+        } else {
+            return 0;
+        }
+    #endif
+}
+
+void pwm_in_init() {
+    GPIO_InitTypeDef GPIO_InitStruct;
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_PULLDOWN;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
+    GPIO_InitStruct.Alternate = GPIO_AF2_TIM5;
+
+    TIM_IC_InitTypeDef sConfigIC;
+    sConfigIC.ICPolarity = TIM_INPUTCHANNELPOLARITY_BOTHEDGE;
+    sConfigIC.ICSelection = TIM_ICSELECTION_DIRECTTI;
+    sConfigIC.ICPrescaler = TIM_ICPSC_DIV1;
+    sConfigIC.ICFilter = 15;
+
+#if HW_VERSION_MAJOR == 3 && HW_VERSION_MINOR >= 3
+    for (int gpio_num = 1; gpio_num <= PWM_IN_COUNT; ++gpio_num) {
+#else
+    int gpio_num = 4; {
+#endif  
+        // Activate timer interrupt if the pin is being used as a PWM input
+        bool pwm_dir_pin = false;
+        for (uint8_t i = 0; i < PWM_IN_COUNT; ++i) {
+            if (board_config.pwm_mappings[i].gpio_direction_pin == gpio_num) {
+                pwm_dir_pin = true;
+            }
+        }
+        if (is_endpoint_ref_valid(board_config.pwm_mappings[gpio_num - 1].endpoint) || pwm_dir_pin) {
+            GPIO_InitStruct.Pin = get_gpio_pin_by_pin(gpio_num);
+            HAL_GPIO_DeInit(get_gpio_port_by_pin(gpio_num), get_gpio_pin_by_pin(gpio_num));
+            HAL_GPIO_Init(get_gpio_port_by_pin(gpio_num), &GPIO_InitStruct);
+            HAL_TIM_IC_ConfigChannel(&htim5, &sConfigIC, gpio_num_to_tim_2_5_channel(gpio_num));
+            HAL_TIM_IC_Start_IT(&htim5, gpio_num_to_tim_2_5_channel(gpio_num));
+        }
```

Comment:
It looks like I did that to handle the case where duty-cycle PWM is at 100% or 0% and the direction pin changes.  Either it needed to be polled or a GPIO interrupt needed to be set up, because there would be no PWM edge to trigger a state check.  It wasn't to handle dual use as PWM and direction input.